"Whops The URL must be of scheme file" error while trying to build the code on windows

# Please view our [Technical Support Guide](https://github.com/microsoft/BotFramework-WebChat/tree/main/docs/TECHNICAL_SUPPORT_GUIDE.md) before filing a new issue.

<!-- ATTENTION: Bot Framework internals, please remove the `customer-reported` and `Bot Services` labels before submitting this issue. -->

<!-- [GitHub issues](https://github.com/microsoft/botframework-webchat/issues) should be used for bugs and feature requests. See the [Technical Support Guide](https://github.com/microsoft/BotFramework-WebChat/tree/main/docs/TECHNICAL_SUPPORT_GUIDE.md) to get support related to Bot Framework and Web Chat. -->

## Screenshots

<!-- If applicable, add screenshots to help explain your problem. -->
<!-- Be sure to remove or obscure personally identifiable information from your code and screenshots -->

## Version

<!-- What version of Web Chat are you using? Are you using the CDN? NPM package? Or embedding Web Chat to your site via `<iframe>`? -->

> To determine what version of Web Chat you are running, open your browser's development tools, and paste the following line of code into the console.
>
> `[].map.call(document.head.querySelectorAll('meta[name^="botframework-"]'), function (meta) { return meta.outerHTML; }).join('\n')`
>
> If you are using Web Chat outside of a browser, please specify your hosting environment. For example, React Native on iOS, Cordova on Android, SharePoint, PowerApps, etc.

## Describe the bug

<!-- Give a clear and concise description of what the bug is. -->
<!-- Please be sure to add screenshots of the console errors in your browser, if there are any -->

## Steps to reproduce

1. Go to '...'
1. Click on '....'
1. Scroll down to '....'
1. See error

## Expected behavior

<!-- Give a clear and concise description of what you expect to happen when following the reproduction steps above. -->

## Additional context

<!-- Add any other context about the problem here.-->

[Bug]
TypeError [ERR_INVALID_URL_SCHEME]: The URL must be of scheme file
    at mkdirSync (node:fs:1318:10)
    at default (file:///D:/opensource/BotFramework-WebChat/packages/support/cldr-data-downloader/src/index.js:50:5)
    at node:internal/util:335:7
    at new Promise (<anonymous>)
    at default (node:internal/util:321:12)
    at file:///D:/opensource/BotFramework-WebChat/packages/support/cldr-data/src/install.mjs:133:11 {
  code: 'ERR_INVALID_URL_SCHEME'

Getting this error while trying to build code in my windows desktop.(Windows 10 Pro)

I was able to repro (sort of) as I didn't get that same error. But, running `npm bootstrap` did fail for me. Working from a clean clone of the repo, my attempt failed an integrity check.

```
> botframework-webchat-root@4.14.2-0 bootstrap
> lerna bootstrap --ci

lerna notice cli v4.0.0
lerna info ci enabled
lerna info Bootstrapping 13 packages
lerna info Installing external dependencies
lerna ERR! npm ci --legacy-peer-deps exited 1 in 'botframework-webchat-core'
lerna ERR! npm ci --legacy-peer-deps stderr:
npm WARN old lockfile
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile
npm WARN old lockfile Error: Integrity checksum failed when using sha512: wanted sha512-R1PXiz31Uc0Vxy4OEOm07x0oSjKAdPPCh3tPivn/Eo8cvz6gveAeuyUUPB21Hoiif0uoPQSSdhIPS3352nvdyQ== but got sha512-hprxVPu6e5Kdp2puZUmvOGjaLv9TCe58E/Fl6hRq4YiVQxIcNvuq6uTM2r1mT/oPskuS9CgR+I94sqAYv0NGKA==.
npm WARN old lockfile     at C:\Program Files\nodejs\node_modules\npm\node_modules\pacote\lib\registry.js:145:39
npm WARN old lockfile     at async Array.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:741:9)
npm WARN old lockfile  Could not fetch metadata for @babel/helper-split-export-declaration@7.14.5 Error: Integrity checksum failed when using sha512: wanted sha512-R1PXiz31Uc0Vxy4OEOm07x0oSjKAdPPCh3tPivn/Eo8cvz6gveAeuyUUPB21Hoiif0uoPQSSdhIPS3352nvdyQ== but got sha512-hprxVPu6e5Kdp2puZUmvOGjaLv9TCe58E/Fl6hRq4YiVQxIcNvuq6uTM2r1mT/oPskuS9CgR+I94sqAYv0NGKA==.
npm WARN old lockfile     at C:\Program Files\nodejs\node_modules\npm\node_modules\pacote\lib\registry.js:145:39
npm WARN old lockfile     at async Array.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:741:9) {
npm WARN old lockfile   code: 'EINTEGRITY'
npm WARN old lockfile }
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
npm ERR! code 1
npm ERR! path C:\Users\Steven Kanberg\source\WebChat\BotFramework-WebChat\packages\core\node_modules\p-defer-es5
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c npm run postinstall:commonjs && npm run postinstall:esm
npm ERR! > p-defer-es5@2.0.0 postinstall:commonjs
npm ERR! > node ./scripts/esbuild.mjs p-defer ./src/index.js cjs ./lib/index.bundled.js && babel ./lib/index.bundled.js --config-file ./babel.config.commonjs.json --out-file ./lib/index.js
npm ERR! node:events:368
npm ERR!       throw er; // Unhandled 'error' event
npm ERR!       ^
npm ERR!
npm ERR! Error: spawn C:\Users\Steven Kanberg\source\WebChat\BotFramework-WebChat\packages\core\node_modules\esbuild\esbuild.exe ENOENT
npm ERR!     at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
npm ERR!     at onErrorNT (node:internal/child_process:477:16)
npm ERR!     at processTicksAndRejections (node:internal/process/task_queues:83:21)
npm ERR! Emitted 'error' event on ChildProcess instance at:
npm ERR!     at Process.ChildProcess._handle.onexit (node:internal/child_process:288:12)
npm ERR!     at onErrorNT (node:internal/child_process:477:16)
npm ERR!     at processTicksAndRejections (node:internal/process/task_queues:83:21) {
npm ERR!   errno: -4058,
npm ERR!   code: 'ENOENT',
npm ERR!   syscall: 'spawn C:\\Users\\Steven Kanberg\\source\\WebChat\\BotFramework-WebChat\\packages\\core\\node_modules\\esbuild\\esbuild.exe',
npm ERR!   path: 'C:\\Users\\Steven Kanberg\\source\\WebChat\\BotFramework-WebChat\\packages\\core\\node_modules\\esbuild\\esbuild.exe',
npm ERR!   spawnargs: [ '--service=0.12.19', '--ping' ]
npm ERR! }

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Steven~1\AppData\Roaming\npm-cache\_logs\2022-01-05T20_27_11_053Z-debug.log
lerna ERR! npm ci --legacy-peer-deps exited 1 in 'botframework-webchat-core'
lerna WARN complete Waiting for 7 child processes to exit. CTRL-C to exit immediately.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Fixed
+
+-  Fixes [#4102](https://github.com/microsoft/BotFramework-WebChat/issues/4102). Fixed `cldr-data-downloader` package not working properly on Windows, by [@compulim](https://github.com/compulim) in PR [#4223](https://github.com/microsoft/BotFramework-WebChat/pull/4223)
+
 ### Changed
 
 -  Bumped all dependencies to the latest versions, by [@compulim](https://github.com/compulim) in PR [#4195](https://github.com/microsoft/BotFramework-WebChat/pull/4195)

--- a/packages/support/cldr-data-downloader/package.json
+++ b/packages/support/cldr-data-downloader/package.json
@@ -44,7 +44,7 @@
     "bump:prod": "npm install $(cat package.json | jq -r '(.dependencies | keys) - .skipBump | .[]' | awk '{print $1 \"@latest\"}')",
     "eslint": "npm run precommit",
     "precommit": "npm run precommit:eslint -- src",
-    "precommit:eslint": "../../../node_modules/.bin/eslint --report-unused-disable-directives --max-warnings 0"
+    "precommit:eslint": "node ../../../node_modules/eslint/bin/eslint.js --report-unused-disable-directives --max-warnings 0"
   },
   "skipBump": [],
   "dependencies": {

--- a/packages/support/cldr-data-downloader/src/index.js
+++ b/packages/support/cldr-data-downloader/src/index.js
@@ -9,7 +9,9 @@
 
 'use strict';
 
+import { fileURLToPath } from 'url';
 import { mkdirSync } from 'fs';
+import { resolve } from 'path';
 import assert from 'assert';
 
 import { isUrl, readJSON } from './util.js';
@@ -37,17 +39,20 @@ export default function (srcUrl, destPath, options, callback) {
     options = {};
   }
 
-  assert(typeof srcUrl === 'string', 'must include srcUrl (e.g., "http://www.unicode.org/Public/cldr/26/json.zip")');
+  assert(typeof srcUrl === 'string', 'must include srcUrl (e.g., "https://www.unicode.org/Public/cldr/26/json.zip")');
 
   assert(typeof destPath === 'string', 'must include destPath (e.g., "./cldr")');
   assert(!/\.\./u.test(destPath), '"destPath" must not contains "..".');
+  assert(!/^file:/u.test(destPath), '"destPath" must not be URL.');
 
   assert(typeof options === 'object', 'invalid options');
 
   assert(typeof callback === 'function', 'must include callback function');
 
+  destPath = resolve(fileURLToPath(import.meta.url), destPath);
+
   try {
-    mkdirSync(new URL(destPath, import.meta.url), { recursive: true });
+    mkdirSync(destPath, { recursive: true });
   } catch (err) {
     if (err.code !== 'EEXIST') {
       throw err;

--- a/packages/support/cldr-data-downloader/src/util.js
+++ b/packages/support/cldr-data-downloader/src/util.js
@@ -28,10 +28,9 @@ function deepEqual(a, b) {
 
 function isUrl(urlOrPath) {
   try {
-    // eslint-disable-next-line no-new
-    new URL(urlOrPath);
+    const url = new URL(urlOrPath);
 
-    return true;
+    return ['file:', 'http:', 'https:'].some(protocol => url.protocol === protocol);
   } catch (err) {
     return false;
   }

--- a/packages/support/cldr-data/package.json
+++ b/packages/support/cldr-data/package.json
@@ -44,7 +44,7 @@
     "eslint": "npm run precommit",
     "install": "node ./src/install.mjs",
     "precommit": "npm run precommit:eslint -- src",
-    "precommit:eslint": "../../../node_modules/.bin/eslint --report-unused-disable-directives --max-warnings 0"
+    "precommit:eslint": "node ../../../node_modules/eslint/bin/eslint.js --report-unused-disable-directives --max-warnings 0"
   },
   "skipBump": [],
   "dependencies": {

--- a/packages/support/cldr-data/src/install.mjs
+++ b/packages/support/cldr-data/src/install.mjs
@@ -90,7 +90,7 @@ let isNpm3;
 
   if (process.env.CLDR_URL) {
     console.warn('CLDR_URL is deprecated, use CLDR_DATA_URLS_JSON instead.');
-    srcUrl = srcUrl.replace('http://www.unicode.org/Public/cldr', process.env.CLDR_URL.replace(/\/$/u, ''));
+    srcUrl = srcUrl.replace('https://www.unicode.org/Public/cldr', process.env.CLDR_URL.replace(/\/$/u, ''));
   } else {
     if (process.env.CLDR_DATA_URLS_JSON) {
       srcUrl = process.env.CLDR_DATA_URLS_JSON;

